crates/oxc_span/src/ident.rs

```diff
@@ -0,0 +1,384 @@
+#![expect(
+    missing_docs,
+    clippy::missing_panics_doc,
+    clippy::cast_possible_truncation,
+    clippy::len_without_is_empty,
+    clippy::undocumented_unsafe_blocks,
+    clippy::cast_lossless
+)] // TODO
+use oxc_allocator::{Allocator, CloneIn, Dummy, FromIn};
+
+use rustc_hash::FxHasher;
+
+use std::{
+    borrow::Cow,
+    hash::{BuildHasher, Hash, Hasher},
+    marker::PhantomData,
+    ops::Deref,
+    ptr::NonNull,
+};
+
+#[cfg(feature = "serialize")]
+use oxc_estree::{ESTree, Serializer as ESTreeSerializer};
+#[cfg(feature = "serialize")]
+use serde::{Serialize, Serializer as SerdeSerializer};
+
+use crate::{Atom, CompactStr, ContentEq};
+
+#[derive(Clone, Copy, Eq)]
+pub struct Ident<'a> {
+    ptr: NonNull<u8>,
+    // Length in bottom 32 bits, hash in top 32 bits
+    len_and_hash: u64,
+    _marker: PhantomData<&'a str>,
+}
+
+impl<'a> Ident<'a> {
+    pub fn new(s: &'a str) -> Self {
+        let ptr = NonNull::from(s).cast::<u8>();
+
+        // Produce a hash of the string
+        // NOTE: This is creating a fresh hasher for each identifier, which is probably bad for performance?
+        // But, I want to see how terrible it is and keep the API simple for testing.
+        let hash = {
+            let mut hasher = FxHasher::default();
+            s.hash(&mut hasher);
+            hasher.finish()
+        };
+
+        // With FxHasher, highest entropy is in top 32 bits. Clear bottom 32 bits.
+        let hash = hash & !(u32::MAX as u64);
+        // We know `s.len()` is <= u32::MAX so don't bother masking it
+        let len = s.len() as u64;
+
+        // Identifiers cannot have zero size.
+        assert!(len > 0, "identifiers cannot have zero length");
```

Comment:
For purposes of benchmarking, I'd suggest commenting out this assertion, or making it `debug_assert!`. We'll find ways to guarantee `len > 0` and `len <= u32::MAX` at call sites, so eventually we won't need this assertion and it'll still be safe.

Reply:
Odd. We should investigate why this causes a fail on conformance. But that's for another time...